java/src/main/java/io/opentracing/grpc/ServerTracingInterceptor.java

```diff
@@ -0,0 +1,234 @@
+        ServerCall.Listener<ReqT> listenerWithContext = Contexts
+            .interceptCall(ctxWithSpan, call, headers, next);
+
+        ServerCall.Listener<ReqT> tracingListenerWithContext = 
+            new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(listenerWithContext) {
+
+            @Override
+            public void onMessage(ReqT message) {
+                if (streaming || verbose) { span.log("Message received", message); }
+                delegate().onMessage(message);
+            }
+
+            @Override 
+            public void onHalfClose() {
+                if (streaming) { span.log("Client finished sending messages", null); }
+                delegate().onHalfClose();
+            }
+
+            @Override
+            public void onCancel() {
+                span.log("Call cancelled", null);
+                span.finish();
+                delegate().onCancel();
+            }
+
+            @Override
+            public void onComplete() {
+                if (verbose) { span.log("Call completed", null); }
+                span.finish();
+                delegate().onComplete();
+            }
+        };
+
+        return tracingListenerWithContext; 
+    }
+
+    private Span getSpanFromHeaders(Map<String, String> headers, String operationName) {
+        Span span;
+        try {
+            SpanContext parentSpanCtx = tracer.extract(Format.Builtin.HTTP_HEADERS, 
+                new TextMapExtractAdapter(headers));
+            if (parentSpanCtx == null) {
+                span = tracer.buildSpan(operationName).start();
+            } else {
+                span = tracer.buildSpan(operationName).asChildOf(parentSpanCtx).start();
+            }
+        } catch (IllegalArgumentException iae){
+            span = tracer.buildSpan(operationName)
+                .withTag("Error", "Extract failed and an IllegalArgumentException was thrown")
+                .start();
+        }
+        return span;  
+    }
+
+    /**
+     * Builds the configuration of a ServerTracingInterceptor.
+     */
+    public static class Builder {
+        private final Tracer tracer;
+        private String operationName;
+        private boolean streaming;
+        private boolean verbose;
+        private Set<ServerRequestAttribute> tracedAttributes;
+
+        /**
+         * @param tracer to use for this interceptor
+         * @return a Builder with default configuration
+         */
+        public Builder(Tracer tracer) {
+            this.tracer = tracer;
+            this.operationName = "";
+            this.streaming = false;
+            this.verbose = false;
+            this.tracedAttributes = new HashSet<ServerRequestAttribute>();
+        }
+
+        /**
+         * @param operationName for all spans created by this interceptor
+         * @return this Builder with configured operation name
+         */
+        public Builder withOperationName(String operationName) {
```

Comment:
(and if gRPC requires Java 8 or whatever, I think (??) there's a more concise way to say something like the above???)